ramanujan/enumerators/FREnumerator.py

```diff
@@ -0,0 +1,123 @@
+import math
+import mpmath
+
+from .RelativeGCFEnumerator import RelativeGCFEnumerator
+from collections import namedtuple
+
+CONVERGENCE_THRESHOLD = 0.1
+BURST_NUMBER = 200
+FIRST_ENUMERATION_MAX_DEPT = 5_000
+MIN_ITERS = 1
+
+Match = namedtuple('Match', 'rhs_an_poly rhs_bn_poly')
+RefinedMatch = namedtuple('RefinedMatch', 'rhs_an_poly rhs_bn_poly val c_top c_bot precision')
+
+
+def check_for_fr(an_iterator, bn_iterator, an_deg, burst_number=BURST_NUMBER, min_iters=MIN_ITERS):
+    """
+    As the calculation for p and q goes on, the GCD for the two grows. 
+    We've noticed that conjectures tends to have a GCD that grows in a super exponential manner (we call that Factorial
+    Reduction).
+    This function test if a GCF has factorial reduction.
+    """
+    calculated_values = []
+    num_of_calculated_vals = 0
+
+    prev_q = 0
+    q = 1
+    prev_p = 1
+    # This is a ugly hack but it works. a[0] is handled before the rest here:
+    p = an_iterator.__next__()  # will place a[0] to p
+    bn_iterator.__next__()  # b0 is discarded
+
+    next_gcd_calculation = burst_number if burst_number >= min_iters else min_iters
+
+    for i, (a_i, b_i) in enumerate(zip(an_iterator, bn_iterator)):
+        tmp_a = q
+        tmp_b = p
+
+        q = a_i * q + b_i * prev_q
+        p = a_i * p + b_i * prev_p
+
+        prev_q = tmp_a
+        prev_p = tmp_b
+
+        if i == next_gcd_calculation:
+            num_of_calculated_vals += 1
+            next_gcd_calculation += burst_number
+
+            calculated_values.append(
+                mpmath.log(mpmath.mpf(math.gcd(p, q))) / mpmath.mpf(i) +
+                an_deg * (-mpmath.log(i) + 1)
+            )
+
+            # This test fails for GCF without FR. Checking it early allows us do discard a lot of GCFs
```

Comment:
"This test" is not a good way to describe code. It would be better to describe the test, such as: "The GCF should grow super exponentially at this point to have a FR.".

Reply:
Relative enumerator has a great comment that could be copied here as well.